ops/testing.py

```diff
@@ -547,10 +602,12 @@ def detach_storage(self, storage_id: str) -> None:
             raise RuntimeError('cannot detach storage before Harness is initialised')
         storage_name, storage_index = storage_id.split('/', 1)
         storage_index = int(storage_index)
-        if self._backend._storage_is_attached(storage_name, storage_index) and self._hooks_enabled:
-            self.charm.on[storage_name].storage_detaching.emit(
-                model.Storage(storage_name, storage_index, self._backend))
-        self._backend._storage_detach(storage_id)
+        storage_attached = self._backend._storage_is_attached(  # pyright:ReportPrivateUsage=false
+            storage_name, storage_index)
+        if storage_attached and self._hooks_enabled:
+            self.charm.on[storage_name].storage_detaching.emit(  # type: ignore
+                model.Storage(storage_name, storage_index, self._backend))  # type: ignore
```

Comment:
These and all the several other type ignore comments can be removed with that fix I mentioned above.